### ISSUE TITLE
Frozen hashes breaking truncator functionality

### ISSUE THREAD

**Issue by bengreville** [airbrake/airbrake-ruby#281]
# Issue discovered in Truncator.rb
## Summary

Frozen hashes are unable to be truncated correctly by 
https://github.com/airbrake/airbrake-ruby/blob/master/lib/airbrake-ruby/truncator.rb#L103.

![truncation error](https://user-images.githubusercontent.com/1417959/32422965-740d9c32-c308-11e7-919f-4400a330b180.png)

## Description

Currently when truncating hashes for logging purposes, the original hash values get modified.

While immutability would be preferred, reusing the hash is likely here for keeping memory usage down. This is however creating an error when a frozen hash is passed in.

## Potential Solution 👍 

This change resolves this by checking for a frozen hash, and duplicates it allowing for modification when required.

There is a pull request https://github.com/airbrake/airbrake-ruby/pull/280 (currently closed) outlining what a fix might look like, with related rspec tests included. 

**However there may be further mutations occurring upstream that I'm not aware of, which may make this a breaking change.**

Looking for input from the wider *airbrake-ruby developer community* to ensure the correct solution is found.

**Comment by kyrylo** [comment]
Can you please outline what possible problems do you see with the approach in #280? It wasn't clear to me. Maybe a concrete example?

**Comment by pond** [comment]
@kyrylo:

In itself, the PR is fine and is a simple way to stop Hash truncation from modifying something that's frozen. The problem is that in wider use, it'd break Airbrake Ruby. (Aside: We also noted in passing that even something like a frozen user-generated Exception would cause Airbrake Ruby to break, because it tries to call `set_backtrace` on the given instance; so there is a wider problem of undocumented input parameter mutation within the gem anyway and, IMHO, the class of all such "modifies input parameter" problems ought to be addressed consistently in one PR).

The most minor issue arises from the way in which `truncate_hash` is called - the top level `truncate_object`:

```ruby
    def truncate_object(object, seen = {})
      return seen[object] if seen[object]

      seen[object] = '[Circular]'.freeze
      truncated =
        if object.is_a?(Hash)
          truncate_hash(object, seen)
        elsif object.is_a?(Array)
          truncate_array(object, seen)
        elsif object.is_a?(Set)
          truncate_set(object, seen)
        else
          raise Airbrake::Error,
                "cannot truncate object: #{object} (#{object.class})"
        end
      seen[object] = truncated
    end
```

In https://github.com/airbrake/airbrake-ruby/pull/280/files we see that a frozen Hash would be `dup`'d, leading to the value in `truncated` and, therefore, the value in `seen[object]` being set to the new, duplicated object. This breaks the cache; if we tried to re-truncate the same frozen Hash again, the `seen` cache wouldn't have it in - it would only have recorded the duplicate - so it'd repeat the whole truncation and "cache" another duplication.

Higher level, things get worse. There's a rather distressing design pattern within Airbrake Ruby wherein the code often doesn't use return values on methods, instead assuming called objects have modified state. This is true of the use of the entire `Airbrake::Truncator` class. When e.g. `notify_sync` is called, the Airbrake notice gets compiled for HTTPS `POST` via `to_json` sent to `Airbrake::Notice`:

```ruby
    def to_json
      loop do
        begin
          json = @payload.to_json
        rescue *JSON_EXCEPTIONS => ex
          @config.logger.debug("#{LOG_LABEL} `notice.to_json` failed: #{ex.class}: #{ex}")
        else
          return json if json && json.bytesize <= MAX_NOTICE_SIZE
        end

        break if truncate == 0
      end
    end
```

There's a rather nasty idiom in `break if truncate == 0` - that's actually a method call to:

```ruby
    def truncate
      TRUNCATABLE_KEYS.each { |key| @truncator.truncate_object(self[key]) }
      ...
```

This is a bit obscure - the code isn't very clean IMHO, it's kinda hard to decode - it's leaning on a bespoke `Notice#[]` method that's actually taking keys out of the `@payload` instance variable we saw earlier. So this is just calling `truncate_object` on `@payload` fields _AND IGNORING THE RETURN VALUE_. It's expecting them to be truncated inline.

So TL;DR, when the bottom level Truncate mechanism starts truncating a duplicated Hash, then even though it's returning the duplicate, the caller is ignoring the return value and making an unwarranted assumption that the parameter was actually mutated in place. It wouldn't be, so we'd go to all the trouble of duplicating and truncating a Hash, only to then send the un-truncated original one to Airbrake, With Hilarious Consequences.

**Comment by pond** [comment]
The reason we closed #280 was the realisation that with problems so pervasive, it'd require a large PR and given that we're a commercial outfit we'd have to find budget / time / justification for that if we were to do it ourselves. Method `truncate_object`'s documentation says "This method is public to simplify testing. You probably want to use {truncate_notice} instead", but that's outdated as no such method as `truncate_notice` exists; there's a pervasive pattern of near-invisible mutation of passed objects, with even `truncate_object` saying it has no return value ("@return [void]"); we'd want to tidy that up; one can even spot efficiencies that would be nice to incorporate. If we decide that in low memory conditions all the other things going on in Ruby probably mean you can't Airbrake either way, then we could just decide to _always_ use a duplicate a Hash for truncation. So rather than this:

```ruby
def truncate_hash(hash, seen)
  hash.each_with_index do |(key, val), idx|
    if idx < @max_size
      hash[key] = truncate(val, seen)
    else
      hash.delete(key)
    end
  end
end
```

...which _must_ go through every single top-level key and recursively truncate the values _or_ manually delete every single remaining key past the size limit, we could do something cleaner (UNTESTED / PSEUDOCODE EXAMPLE):

```ruby
def truncate_hash(hash, seen)
  truncated_hash = {}

  hash.each_with_index do |(key, val), idx|
    break if idx >= @max_size
    truncated_hash[key] = truncate(val, seen)
  end

  return truncated_hash
end
```

...this exits early as soon as it reaches the count limit and doesn't need to duplicate the _entire_ input Hash for the frozen case, as per #280. It only writes as many keys as it needs.

**Comment by kyrylo** [comment]
@pond thanks for the detailed write-up. I rewrote the class to not mutate objects (see: https://github.com/airbrake/airbrake-ruby/pull/283). Can you please give it a spin and let me know what you think?

**Comment by pond** [comment]
@kyrylo Nice! Feedback in https://github.com/airbrake/airbrake-ruby/pull/283#pullrequestreview-75245084 :-)

**Comment by kyrylo** [comment]
@bengreville @pond I just released airbrake-ruby v2.6.0 with the fix. Thanks a bunch for helping!

**Comment by bengreville** [comment]
Thanks @kyrylo for the fast response! Greatly appreciated.

**Comment by pond** [comment]
Legend! Thanks @kyrylo :)
